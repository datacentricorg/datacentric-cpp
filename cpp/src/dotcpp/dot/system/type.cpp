/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/precompiled.hpp>
#include <dot/implement.hpp>
#include <dot/system/type.hpp>
#include <dot/system/object_impl.hpp>
#include <dot/system/reflection/method_info.hpp>
#include <dot/system/reflection/constructor_info.hpp>
#include <dot/system/collections/generic/list.hpp>
#include <dot/system/string.hpp>

namespace dot
{
    type_builder_impl::type_builder_impl(string name_space, string name, string cpp_name)
        : full_name_(name_space + "." + name)
    {
        type_ = new type_impl(name_space, name);
        type_impl::get_type_map()[full_name_] = type_;
        type_impl::get_type_map()[name] = type_;
        type_impl::get_type_map()[cpp_name] = type_;
    }

    type type_builder_impl::build()
    {
        type_->fill(this);

        // Fill derived types map
        type base_type = base_;
        while (base_type != nullptr)
        {
            auto iter = type_impl::get_derived_types_map().find(base_type->full_name());
            if (iter == type_impl::get_derived_types_map().end())
            {
                iter = type_impl::get_derived_types_map().insert({base_type->full_name(), make_list<type>()}).first;
            }
            else if (iter->second == nullptr)
            {
                iter->second = make_list<type>();
            }
            iter->second->add(type_);
            base_type = base_type->get_base_type();
        }

        return type_;
    }

    void type_impl::fill(const type_builder& data)
    {
        if (!data->base_.is_empty() && !data->base_->get_methods().is_empty() && data->base_->get_methods()->count())
        {
            if (data->methods_.is_empty())
            {
                data->methods_ = make_list<method_info>();
            }

            list<method_info> base_methods = data->base_->get_methods();
            list<method_info> make_methods = make_list<method_info>();
            for (method_info meth_info_data : base_methods)
            {

                make_methods->add(meth_info_data);
            }

            for (method_info meth_info_data : data->methods_)
            {
                make_methods->add(meth_info_data);
            }

            data->methods_ = make_methods;
        }

        if (!data->methods_.is_empty())
        {
            this->methods_ = make_list<method_info>(data->methods_->count());
            int i = 0;
            for (method_info meth_info_data : data->methods_)
            {
                this->methods_[i++] = meth_info_data;
            }
        }
        else
        {
            this->methods_ = make_list<method_info>(0);
        }

        if (!data->ctors_.is_empty())
        {
            this->ctors_ = make_list<constructor_info>(data->ctors_->count());
            int i = 0;
            for (constructor_info ctor_info_data : data->ctors_)
            {
                this->ctors_[i++] = ctor_info_data;
            }
        }
        else
        {
            this->ctors_ = make_list<constructor_info>(0);
        }

        if (!data->base_.is_empty() && !data->base_->get_fields().is_empty() && data->base_->get_fields()->count())
        {
            if (data->fields_.is_empty())
            {
                data->fields_ = make_list<field_info>();
            }

            list<field_info> base_fields = data->base_->get_fields();
            list<field_info> make_fields = make_list<field_info>();
            for (field_info field_info_data : base_fields)
            {

                make_fields->add(field_info_data);
            }

            for (field_info field_info_data : data->fields_)
            {
                make_fields->add(field_info_data);
            }

            data->fields_ = make_fields;
        }

        if (!data->fields_.is_empty())
        {
            this->fields_ = make_list<field_info>(data->fields_->count());
            int i = 0;
            for (field_info ctor_info_data : data->fields_)
            {
                this->fields_[i++] = ctor_info_data;
            }
        }
        else
        {
            this->fields_ = make_list<field_info>(0);
        }

        if (!data->interfaces_.is_empty())
        {
            this->interfaces_ = make_list<type>(data->interfaces_->count());
            int i = 0;
            for (type interface : data->interfaces_)
            {
                this->interfaces_[i++] = interface;
            }
        }
        else
        {
            this->interfaces_ = make_list<type>(0);
        }

        if (!data->generic_args_.is_empty())
        {
            this->generic_args_ = make_list<type>(data->generic_args_->count());
            int i = 0;
            for (type arg : data->generic_args_)
            {
                this->generic_args_[i++] = arg;
            }
        }
        else
        {
            this->generic_args_ = make_list<type>(0);
        }

        this->base_ = data->base_;
        this->is_class = data->is_class_;
        this->is_enum = data->is_enum_;
    }

    type_impl::type_impl(string nspace, string name)
    {
        this->name_space = nspace;
        this->name = name;
    }

    method_info type_impl::get_method(string name)
    {
        if (methods_.is_empty()) return nullptr;

        for (auto method : methods_)
        {
            if (method->name == name)
                return method;
        }

        return nullptr;
    }

    type type_impl::get_interface(string name)
    {
        if (interfaces_.is_empty()) return nullptr;

        for (auto interface : interfaces_)
        {
            if (interface->name == name)
                return interface;
        }

        return nullptr;
    }

    field_info type_impl::get_field(string name)
    {
        if (fields_.is_empty()) return nullptr;

        for (auto field : fields_)
        {
            if (field->name == name)
                return field;
        }

        return nullptr;

    }

    bool type_impl::equals(object obj)
    {
        if (obj.is<type>())
            return this->full_name() == ((type)obj)->full_name();

        return false;
    }

    size_t type_impl::hash_code()
    {
        return this->full_name()->hash_code();
    }
}
