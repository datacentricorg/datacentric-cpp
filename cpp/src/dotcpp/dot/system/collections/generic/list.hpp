/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/system/ptr.hpp>
#include <dot/system/collections/list_base.hpp>

namespace dot
{
    template <class T> class ListImpl; template <class T> using List = Ptr<ListImpl<T>>;
    class TypeBuilderImpl; using TypeBuilder = Ptr<TypeBuilderImpl>;

    /// Represents a strongly typed collection of objects that can be accessed by index.
    template <class T>
    class ListImpl : public virtual ObjectImpl, public std::vector<T>, public ListBaseImpl
    {
        template <class R> friend List<R> make_list();
        template <class R> friend List<R> make_list(const std::vector<R> & obj);
        template <class R> friend List<R> make_list(std::vector<R>&& obj);
        template <class R> friend List<R> make_list(const std::initializer_list<R> & obj);
        template <class R> friend List<R> make_list(int size);

        typedef ListImpl<T> self;
        typedef std::vector<T> base;

    private: // CONSTRUCTORS

        /// Create empty list with default initial capacity.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        ListImpl() {}

        /// Construct from vector using deep copy semantics.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(const std::vector<T>& obj) : base(obj) {}

        /// Construct from vector using move semantics.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(std::vector<T>&& obj) : base(obj) {}

        /// Construct from initializer list.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(const std::initializer_list<T>& obj) : base(obj) {}

        /// Construct from int.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(int size) : base(size) {}

    public: // METHODS

        /// The number of items contained in the list.
        int count() { return this->size(); }

        /// Set total number of elements the internal data structure can hold without resizing.
        void set_capacity(int value) { this->reserve(value); }

        /// Adds an Object to the end of the list.
        void add(const T& item) { this->push_back(item); }

        /// Determines whether an element is in the list.
        virtual bool contains(const T& item)
        {
            return std::find(this->begin(), this->end(), item) != this->end();
        }

        /// Adds the elements of the specified collection to the end of the list.
        void add_range(const std::initializer_list<T>& collection)
        {
            this->insert(this->end(), collection.begin(), collection.end());
        }

        /// Adds the elements of the specified collection to the end of the list.
        void add_range(List<T> collection)
        {
            this->insert(this->end(), collection->begin(), collection->end());
        }

        /// Removes the first occurrence of a specific Object from the list.
        bool remove(const T& item)
        {
            auto iter = std::find(this->begin(), this->end(), item);
            if (iter != this->end())
            {
                this->erase(iter);
                return true;
            }
            return false;
        }

    public: // VIRTUAL METHODS

        /// Get Object from collection by index.
        virtual Object get_item(int index) override
        {
            return this->operator[](index);
        }

        /// Set Object from collection by index.
        virtual void set_item(int index, Object value) override
        {
            this->operator[](index) = (T)value;
        }

        /// Get length of collection.
        virtual int get_length() override
        {
            return this->size();
        }

        /// Add Object to end of collection.
        virtual void add_object(Object item) override
        {
            add((T)item);
        }

    public: // OPERATORS

        /// Gets or sets the element at the specified index (const version).
        virtual const T& operator[](int i) const { return base::operator[](i); }

        /// Gets or sets the element at the specified index (non-const version).
        virtual T& operator[](int i) { return base::operator[](i); }

    public: // REFLECTION

        virtual Type get_type() { return typeof(); }
        static Type typeof();
    };

    /// Create empty list with default initial capacity.
    template <class T>
    List<T> make_list() { return new ListImpl<T>(); }

    /// Construct from vector using deep copy semantics.
    template <class T>
    List<T> make_list(const std::vector<T> & obj) { return new ListImpl<T>(obj); }

    /// Construct from vector using move semantics.
    template <class T>
    List<T> make_list(std::vector<T>&& obj) { return new ListImpl<T>(std::forward<std::vector<T>>(obj)); }

    /// Construct from initializer list.
    template <class T>
    List<T> make_list(const std::initializer_list<T> & obj) { return new ListImpl<T>(obj); }

    /// Construct from int.
    template <class T>
    List<T> make_list(int size) { return new ListImpl<T>(size); }
}
