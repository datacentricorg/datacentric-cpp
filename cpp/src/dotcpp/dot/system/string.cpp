/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/precompiled.hpp>
#define FMT_HEADER_ONLY
#include <fmt/format.h>
#include <dot/implement.hpp>
#include <dot/system/string.hpp>
#include <dot/system/object.hpp>
#include <dot/system/nullable.hpp>
#include <dot/system/type.hpp>

namespace dot
{
    string string::empty = make_string("");

    dot::type string_impl::typeof()
    {
        static dot::type result = []()->dot::type
        {
            dot::type t = dot::make_type_builder<string_impl>("dot", "string")
                ->build();
            return t;
        }();
        return result;
    }

    dot::type string_impl::get_type()
    {
        return typeof();
    }


    bool string_impl::equals(object obj)
    {
        if (this == &(*obj)) return true;

        if (obj.is<string>())
        {
            return *this == *obj.as<string>();
        }

        return false;
    }

    size_t string_impl::hash_code()
    {
        return std::hash<std::string>()(*this);
    }

    string string_impl::to_string()
    {
        return this;
    }

    bool string_impl::ends_with(const string& value)
    {
        int p = length() - value->length();
        if (p >= 0 && substr(p, value->length()) == *value)
            return true;
        return false;
    }

    bool string_impl::starts_with(const string& value)
    {
        int p = length() - value->length();
        if (p >= 0 && substr(0, value->length()) == *value)
            return true;
        return false;
    }

    string string_impl::substring(int start_index)
    {
        return make_string(this->substr(start_index));
    }

    string string_impl::substring(int start_index, int length)
    {
        return make_string(this->substr(start_index, length));
    }

    int string_impl::compare_to(const string& str_b) const
    {
        return this->compare(*str_b);
    }

    int string_impl::index_of_any(list<char> any_of)
    {
        size_t pos = find_first_of(any_of->data(), 0, any_of->size());
        if (pos != std::string::npos)
            return pos;
        return -1;
    }

    string string_impl::remove(int start_index)
    {
        return make_string(*this)->erase(start_index);
    }

    string string_impl::remove(int start_index, int count)
    {
        return make_string(*this)->erase(start_index, count);
    }

    string string_impl::replace(const char old_char, const char new_char) const
    {
        string make_str = *this;
        std::replace(make_str->begin(), make_str->end(), old_char, new_char);
        return make_str;
    }

    list<string> string_impl::split(char separator) const
    {
        list<string> result = make_list<string>();

        std::size_t current, previous = 0;
        current = this->find(separator);
        while (current != std::string::npos) {
            result->add(this->substr(previous, current - previous));
            previous = current + 1;
            current = this->find(separator, previous);
        }
        result->add(this->substr(previous, current - previous));

        return result;
    }

    list<string> string_impl::split(string separator) const
    {
        list<string> result = make_list<string>();

        std::size_t current, previous = 0;
        current = this->find_first_of(*separator);
        while (current != std::string::npos) {
            result->add(this->substr(previous, current - previous));
            previous = current + 1;
            current = this->find_first_of(*separator, previous);
        }
        result->add(this->substr(previous, current - previous));

        return result;
    }

    string string_impl::trim() const
    {
        string result = make_string(*this);

        // Trim end
        result->erase(std::find_if(result->rbegin(), result->rend(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }).base(), result->end());

        // Trim start
        result->erase(result->begin(), std::find_if(result->begin(), result->end(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }));

        return result;
    }

    string string_impl::trim_start() const
    {
        string result = make_string(*this);
        result->erase(result->begin(), std::find_if(result->begin(), result->end(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }));
        return result;
    }

    string string_impl::trim_end() const
    {
        string result = make_string(*this);
        result->erase(std::find_if(result->rbegin(), result->rend(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }).base(), result->end());
        return result;
    }

    bool string_impl::contains(const string& s) const
    {
        return this->find(*s) != std::string::npos;
    }

    string string_impl::to_lower() const
    {
        string result = make_string(*this);
        std::transform(result->begin(), result->end(), result->begin(),
            [](unsigned char c) { return std::tolower(c); });
        return result;
    }

    string string_impl::to_upper() const
    {
        string result = make_string(*this);
        std::transform(result->begin(), result->end(), result->begin(),
            [](unsigned char c) { return std::toupper(c); });
        return result;
    }

    bool string::is_null_or_empty(string value)
    {
        if (value == nullptr || value->empty())
            return true;
        return false;
    }

    bool string::operator==(const object& rhs) const
    {
        // If rhs is null, return false. Otherwise, check if
        // the other object is a string. If yes, compare by value.
        // If no, return false.
        if (rhs == nullptr)
        {
            return false;
        }
        else
        {
            string rhs_str = rhs.as<string>();
            if (rhs_str != nullptr) return operator==(rhs_str);
            else return false;
        }
    }

    std::string string::format_impl(fmt::string_view format_str, fmt::format_args args)
    {
        return fmt::vformat(format_str, args);
    }
}
