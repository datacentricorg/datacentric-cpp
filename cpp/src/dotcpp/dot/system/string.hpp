/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/declare.hpp>
#include <fmt/core.h>
#include <dot/detail/const_string_base.hpp>
#include <dot/system/ptr.hpp>
#include <dot/system/char.hpp>

namespace dot
{
    class string_impl; class string;
    template <class T> class list_impl; template <class T> using list = ptr<list_impl<T>>;
    template <class T> class nullable;
    enum class string_split_options;

    /// Immutable string type.
    ///
    /// The string is encoded internally as UTF-8.
    class DOT_CLASS string_impl : public virtual object_impl, public detail::ConstStringBase
    {
        typedef string_impl self;
        typedef detail::ConstStringBase base;
        friend string make_string(const std::string& rhs);
        friend string make_string(const char* rhs);

    public: // CONSTRUCTORS

        /// Creates an empty string.
        string_impl() {}

        /// Initializes from std::string or string literal.
        ///
        /// This constructor is private. Use make_string(str) function instead.
        string_impl(const std::string& value) : base(value) {}

        /// Create from const char*, null pointer is converted to to empty value.
        string_impl(const char* value) : base(value) {}

        /// Create from a single 8-bit character.
        string_impl(char value) : base(std::to_string(value)) {}

    public: // METHODS

        /// Determines whether this instance and a specified object,
        /// which must also be a string object, have the same value.
        bool equals(object obj) override;

        /// Returns the hash code for this string.
        virtual size_t hash_code() override;

        /// Returns this instance of string; no actual conversion is performed.
        virtual string to_string() override;

        /// Determines whether the end of this
        /// string matches the specified string.
        bool ends_with(const string& value);

        /// Determines whether the beginning of this
        /// string matches the specified string.
        bool starts_with(const string& value);

        /// Retrieves a substring which starts at a specified
        /// character position and continues to the end of the string.
        string substring(int start_index);

        /// Retrieves a substring which starts at the specified
        /// character position and has the specified length.
        string substring(int start_index, int length);

        /// Compares this instance with a specified string object and indicates
        /// whether this instance precedes, follows, or appears in the same position
        /// in the sort order as the specified string.
        int compare_to(const string& str_b) const;

        /// Reports the zero-based index of the first occurrence in this instance of any
        /// character in a specified array of Unicode characters.
        int index_of_any(list<char> any_of);

        /// Returns a new string in which all the characters in the current instance, beginning
        /// at a specified position and continuing through the last position, have been deleted.
        string remove(int start_index);

        /// Returns a new string in which a specified number of characters in the current
        /// instance beginning at a specified position have been deleted.
        string remove(int start_index, int count);

        /// Returns a new string in which all occurrences of a specified string
        /// in the current instance are replaced with another specified string.
        string replace(const string& old_value, const string& new_value) const; // TODO - implement

        /// Returns a new string in which all occurrences of a specified character
        /// in this instance are replaced with another specified character.
        string replace(const char old_char, const char new_char) const;

        /// Returns a string array that contains the substrings of the current string
        /// that are delimited by any of the specified 8-bit characters.
        list<string> split(char separator) const;

        /// Splits a string into substrings that are based on the characters in an array.
        list<string> split(string separator) const;

        /// Returns a string array that contains the substrings in
        /// this string that are delimited by any of the specified strings.
        /// A parameter specifies whether to return empty array elements.
        list<string> split(const list<string>& separator, const string_split_options& options) const; // TODO - implement

        /// Returns a new string in which all leading and trailing occurrences
        /// of a set of specified characters from the current string are removed.
        string trim() const;

        /// Removes all the leading occurrences of a set of characters
        /// specified in an array from the current string.
        string trim_start() const;

        /// Removes all the trailing occurrences of a set of characters
        /// specified in an array from the current string.
        string trim_end() const;

        /// Indicates whether the argument occurs within this string.
        bool contains(const string& s) const;

        /// Returns a copy of this string converted to lowercase.
        string to_lower() const;

        /// Returns a copy of this string converted to uppercase.
        string to_upper() const;

    public: // OPERATORS

        /// Returns a string containing characters from lhs followed by the characters from rhs.
        friend string operator+(const string& lhs, const string& rhs);

        /// Returns a string containing characters from lhs followed by the characters from rhs.
        friend bool operator<(const string& lhs, const string& rhs);

    public: // REFLECTION

        virtual type get_type();
        static type typeof();
    };

    /// Pointer to string_impl that has additional constructors compared to ptr(string)
    class DOT_CLASS string : public ptr<string_impl>
    {
        typedef ptr<string_impl> base;

    public: // CONSTANTS

        /// Empty string.
        static string empty;

    public: // CONSTRUCTORS

        /// Create null string.
        string() : base() {}

        /// Take ownership of raw pointer to template argument type.
        /// This also permits construction from null pointer.
        string(string_impl* p) : base(p) {}

        /// Create from std::string.
        string(const std::string& rhs) : base(new string_impl(rhs)) {}

        /// Create from string literal.
        string(const char* rhs) : base(new string_impl(rhs)) {}

        /// Copy constructor.
        string(const string& rhs) : base(rhs) {}

        /// Supports cast (explicit constructor) from object.
        ///
        /// Error if object does is not a boxed int.
        /// Null object becomes empty nullable int.
        explicit string(const ptr<object_impl>& rhs) : base(rhs) {}

    public: // STATIC

        /// Replaces format entries in the specified string with the string
        /// representation of objects in the argument array.
        template <typename ...Args>
        static string format(const string& format_string, const Args& ...args);

        /// Indicates whether the specified string is null or an string.empty string.
        static bool is_null_or_empty(string value);

    public: // OPERATORS

        /// Case sensitive comparison to std::string.
        bool operator==(const std::string& rhs) const { string_impl& impl = base::operator*(); return impl == rhs; }

        /// Case sensitive comparison to std::string.
        bool operator!=(const std::string& rhs) const { return !operator==(rhs); }

        /// Case sensitive comparison to string literal.
        bool operator==(const char* rhs) const { if (rhs == nullptr) return base::operator==(nullptr); string_impl& impl = base::operator*(); return impl == rhs; }

        /// Case sensitive comparison to string literal.
        bool operator!=(const char* rhs) const { return !operator==(rhs); }

        /// Case sensitive comparison to string literal.
        bool operator==(const ptr<string_impl>& rhs) const { return *this == *rhs; }

        /// Case sensitive comparison to string literal.
        bool operator!=(const ptr<string_impl>& rhs) const { return !operator==(rhs); }

        /// Case sensitive comparison to object.
        bool operator==(const object& rhs) const;

        /// Case sensitive comparison to object.
        bool operator!=(const object& rhs) const { return !operator==(rhs); }

    private: // STATIC

        /// Non-template implementation of string.format.
        static std::string format_impl(fmt::string_view format_str, fmt::format_args args);
    };

    /// Create from std::string or string literal using new
    inline string make_string(const std::string& rhs) { return new string_impl(rhs); }

    /// Create from std::string or string literal using new
    inline string make_string(const char* rhs) { return new string_impl(rhs); }

    /// Returns a string containing characters from lhs followed by the characters from rhs.
    inline string operator+(const string& lhs, const string& rhs) { return make_string(*lhs + *rhs); }

    /// Returns a string containing characters from lhs followed by the characters from rhs.
    inline bool operator<(const string& lhs, const string& rhs) { return *lhs < *rhs; }

    /// Helper class for fmt::format arguments conversion
    template <class T>
    struct format_forward {
        static inline auto convert(const T& t) { return t; }
    };

    /// Helper class for fmt::format arguments conversion
    template<>
    struct format_forward<string> {
        static inline std::string convert(const string& s)
        {
            if (s.is_empty())
                return std::string("");
            else
                return *s;
        }

    };

    /// Helper class for fmt::format arguments conversion
    template<class T>
    struct format_forward<ptr<T>> {
        static inline std::string convert(const ptr<T>& o) { return *o->to_string(); }
    };

    /// Helper class for fmt::format arguments conversion
    template<class T>
    struct format_forward<nullable<T>> {
        static inline T convert(const nullable<T>& o) { return o.value_or_default(); }
    };

    template<typename ...Args>
    string string::format(const string& format_string, const Args& ...args)
    {
        return format_impl(*format_string, fmt::make_format_args(format_forward<Args>::convert(args)...));
    }
}

namespace std
{
    /// Implements hash struct used by STL unordered_map for string.
    template <>
    struct hash<dot::string> : public hash<dot::ptr<dot::string_impl>>
    {};

    /// Implements equal_to struct used by STL unordered_map for string.
    template <>
    struct equal_to<dot::string> : public equal_to<dot::ptr<dot::string_impl>>
    {};
}
