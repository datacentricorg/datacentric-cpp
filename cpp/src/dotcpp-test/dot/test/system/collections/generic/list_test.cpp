/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/test/implement.hpp>
#include <approvals/ApprovalTests.hpp>
#include <approvals/Catch.hpp>
#include <dot/system/collections/generic/list.hpp>
#include <dot/system/string.hpp>
#include <dot/system/type.hpp>

namespace dot
{
    static std::stringstream received;

    /// Print double vector to received on one line in JSON format.
    static void print_list(string name, list<double> v)
    {
        received << "\"" << *name << "\" : [ ";
        for (int i = 0; i < v->count(); ++i)
        {
            if (i > 0) received << ", ";
            received << v[i];
        }
        received << " ]" << std::endl;
    }

    TEST_CASE("smoke")
    {
        list<double> a = make_list<double>();
        a->add(0.0);
        a->add(1.0);
        a->add(2.0);

        REQUIRE(a->count() == 3);

    }

    TEST_CASE("interfaces")
    {
        list<double> a = make_list<double>();

        // Add elements to list interface
        list<double> b = a;
        b->add(2.0);
        b->add(1.0);
        b->add(0.0);

        // Check size of the original class and the interface
        REQUIRE(a->count() == 3);
        REQUIRE(b->count() == 3);

        // Access the underlying std::vector<double> class
        print_list("unsorted", a);
        std::sort(a->begin(), a->end());
        print_list("sorted", a);

        // Access by object
        object obj = b;
        // TODO - uncomment next line when output matches C#
        // received << obj->to_string() << std::endl;

        Approvals::verify(received.str());
        received.clear();
    }

    TEST_CASE("iterators")
    {
        list<string> string_list = make_list<string>();
        string_list->add("000");
        string_list->add("111");
        string_list->add("222");
        REQUIRE(string_list->count() == 3);

        int i = 0;
        for (string str : string_list)
        {
            REQUIRE(string_list[i++] == str);
        }
    }

    TEST_CASE("capacity")
    {
        list<string> string_list = make_list<string>();
        string_list->set_capacity(100);
        REQUIRE(string_list->capacity() == 100);
    }

    TEST_CASE("find")
    {
        dot::list<string> string_list = make_list<string>();
        string_list->add("000");
        string_list->add("111");
        string_list->add("222");
        string_list->add("333");
        string_list->add("444");

        // TODO string_list.find_last([](std::string const& s) { return s == "222"; }) = "57";
        // TODO REQUIRE(string_list.find_last_index([](std::string const& s) { return s == "111"; }) == 0);
    }

    TEST_CASE("methods")
    {
        // Add range
        list<int> int_list = make_list<int>();

        int_list->add_range({1, 2});
        REQUIRE(int_list->count() == 2);
        REQUIRE(int_list[0] == 1);
        REQUIRE(int_list[1] == 2);

        int_list->add_range(make_list<int>({ 3, 4 }));
        REQUIRE(int_list->count() == 4);
        REQUIRE(int_list[0] == 1);
        REQUIRE(int_list[1] == 2);
        REQUIRE(int_list[2] == 3);
        REQUIRE(int_list[3] == 4);
    }
}
