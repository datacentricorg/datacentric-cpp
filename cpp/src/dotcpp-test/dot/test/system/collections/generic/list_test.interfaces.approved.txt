"unsorted" : [ 2, 1, 0 ]
"sorted" : [ 0, 1, 2 ]

